dkg-gadget/src/worker.rs

```diff
@@ -1386,84 +1427,96 @@ where
-			.filter_map(|_| futures::future::ready(keygen_gossip_engine.peek_last_message()));
+			.filter_map(move |_| futures::future::ready(keygen_gossip_engine.peek_last_message()));
 		let mut signing_stream = signing_gossip_engine
 			.message_available_notification()
-			.filter_map(|_| futures::future::ready(signing_gossip_engine.peek_last_message()));
-
-		loop {
-			tokio::select! {
-				// This will make the select a bit biased towards the keygen stream and signing
-				// stream.
-				biased;
-				keygen_msg = keygen_stream.next().fuse() => {
-					if let Some(msg) = keygen_msg {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle keygen message for round {}", msg.msg.round_id);
-						match self.process_incoming_dkg_message(msg) {
-							Ok(_) => {
-								self.keygen_gossip_engine.acknowledge_last_message();
-							},
-							Err(e) => {
-								log::error!(target: "dkg_gadget::worker", "Error processing keygen message: {:?}", e);
-							},
-						}
-					}
-				},
-				signing_msg = signing_stream.next().fuse() => {
-					if let Some(msg) = signing_msg {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle signing message for round {}", msg.msg.round_id);
-						match self.process_incoming_dkg_message(msg) {
-							Ok(_) => {
-								self.signing_gossip_engine.acknowledge_last_message();
-							},
-							Err(e) => {
-								log::error!(target: "dkg_gadget::worker", "Error processing signing message: {:?}", e);
-							},
-						}
-					}
-				},
-				notification = self.import_notifications.next().fuse() => {
-					if let Some(notification) = notification {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle Import notification");
-						self.handle_import_notification(notification);
-					} else {
-						log::error!(target: "dkg_gadget::worker", "Import notification stream closed");
-						break;
-					}
-				},
-				notification = self.finality_notifications.next().fuse() => {
-					if let Some(notification) = notification {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle Finality notification");
-						self.handle_finality_notification(notification);
-					} else {
-						log::error!(target: "dkg_gadget::worker", "Finality notification stream closed");
-						break;
-					}
-				},
-				misbehaviour = misbehaviour_rx.recv().fuse() => {
-					if let Some(misbehaviour) = misbehaviour {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle Misbehaviour");
-						gossip_misbehaviour_report(&mut self, misbehaviour);
-					}
-				},
-				error = error_handler_rx.recv().fuse() => {
-					if let Some(error) = error {
-						log::debug!(target: "dkg_gadget::worker", "Going to handle Error");
-						self.handle_dkg_error(error);
-					}
-				},
+			.filter_map(move |_| futures::future::ready(signing_gossip_engine.peek_last_message()));
+
+		// create a task for each of the following tasks:
+		// 1. handle import notifications
+		// 2. handle finality notifications
```

Comment:
Can we remove the finality notifications now actually? They're not necessary

Reply:
Fewer tasks the better, especially for ones we know we don't need.

Reply:
I will try to remove it and see if things are working normally.